File: problems/Bulls and Cows/getHint.js
/**
 * @param {string} secret
 * @param {string} guess
 * @return {string}
 */
var getHint = function(secret, guess) {
    const n = secret.length;
    const keepSecret = Array(10).fill(0);
    let bulls = 0;
    let cows = 0;

    for (let i = 0; n > i; ++i) {
        if(secret[i] === guess[i]) {
            ++bulls;
        } else {
            ++keepSecret[+secret[i]];
        }
    }

    for (let i = 0; n > i; ++i) {
        if (keepSecret[+guess[i]] && secret[i] !== guess[i]) {
            --keepSecret[+guess[i]];
            ++cows;
        }
    }

    return `${bulls}A${cows}B`;
};

File: problems/Bulls and Cows/getHint.java
class Solution {
    public String getHint(String secret, String guess) {
        int n = secret.length();
        int bulls = 0;
        int cows = 0;
        int[] guessStore = new int[10];
        int[] secretStore = new int[10];

        for (int idx = 0; n > idx; ++idx) {
            ++guessStore[guess.charAt(idx) - '0'];
            ++secretStore[secret.charAt(idx) - '0'];

            if (secret.charAt(idx) == guess.charAt(idx)) {
                ++bulls;
            }
        }

        for (int idx = 0; 10 > idx; ++idx) {
            cows += Math.min(guessStore[idx], secretStore[idx]);
        }

        return String.format("%sA%sB", bulls, (cows - bulls));
    }
}
